docs/discovery-naming-conventions.md

```diff
@@ -0,0 +1,126 @@
+
+### Handling names that contain invalid characters
+
+Backend systems other than Kubernetes (e.g. OpenStack) might support names that
+contain characters that are invalid according to the Kubernetes Service naming
+requirements. In these scenarios, discoverers will skip the service and inform
+the user via logs and metrics that the backend service name is not compatible
+with Gimbal.
+
+Even thought this might affect the user experience and cause friction, Gimbal
+prefers to be predictable and safe when it comes to managing backend services
+and endpoints, instead of using "magic" (replacing or removing characters, etc)
+to massage invalid names into valid ones.
+
+### Handling Discovered Names that are longer than 63 characters
+
+When the _Discovered Name_ is longer than 63 characters, it is shortened using
+the following process:
+
+1. Each component of the _Discovered Name_ gets the same number of maximum
+   characters allowed. The _Discovered Name_ has a total of two components, and
+   thus each component is allocated a maximum of 31 characters (one of the 63 is
+   used for the separator).
+
+2. Starting at the last _Discovered Name Component_, check whether it is longer
+   than the allocated number of characters. If it is, take the SHA256 hash of
+   the _Component_. Truncate the excess characters of the _Component_, and
+   replace with the first 6 characters of the hash.
+
+3. If the resulting _Discovered Name_ is still longer than 63 characters, move
+   onto the next _Component_ and shorten using the approach described above.
+
+4. The shortening process produces the final _Discovered Name_.
+
+#### Example
+
+- `${backend-name}`: `us-east-cluster`
+- `${service-name}`: `the-really-long-kube-service-name-that-is-exactly-63-characters`
+
+1. The _Discovered Name_ has a total of two components. Thus, allocate 31
+   characters to each component (62/2, as one char is allocated for the
+   separator).
+
+2. Check if the last _Component_ of the _Discovered Name_ goes over 21 characters:
+
+    ```
+    "the-really-long-kube-service-name-that-is-exactly-63-characters" has 63 characters.
+    ```
+
+3. Shorten the last _Component_ as it is longer than 21 characters:
+
+  ```
+  Take SHA256 of "the-really-long-kube-service-name-that-is-exactly-63-characters"
+  SHA256 hash = 1feeec450b150bcfd731a5e7399890e6c61088fec088eb61f83baa75ca3bd2d9
+  Short hash = 1feeec
+
+  Result = "the-really-long-kube-serv1feeec"
+  ```
+
+4. Check if the resulting _Discovered Name_ is shorter than 63 characters:
+
+    ```
+    "us-east-cluster-the-really-long-kube-serv1feeec" has 47 characters.
+    Thus, we have arrived at our shortened Discovered Name.
+    ```
+
+Discovered name after shortening: `us-east-cluster-the-really-long-kube-serv1feeec`
+
+## Discoverer Specifics
+
+The specifics of each discoverer are documented below.
+
+### Kubernetes discoverer
+
+- `${backend-name}`: The value of the `--cluster-name` flag provided to the
```

Comment:
Do we want to standardize the names? So change the flag to `--backend-name`?

Reply:
That makes sense to me. We can be explicit in the flag's documentation how this value is used. While we are at it, should we also standardize on "backend" vs "upstream"? Backend is less confusing for me, but not sure what's the more common term.

Reply:
Sounds good. I'll open up another issue.